core/shared/src/main/scala/scalaz/zio/stm/STM.scala

```diff
@@ -395,77 +408,114 @@ object STM {
     /**
      * Executes the todos in the current thread, sequentially.
      */
-    final def execTodos(todos: MutableMap[Long, Todo]): Unit = {
-      val it = todos.entrySet().iterator()
-      while (it.hasNext) it.next().getValue().apply()
-    }
+    final def execTodos(todos: MutableMap[TxnId, Todo]): Unit =
+      todos.forEach((_, value) => value.apply())
 
     /**
      * For the given transaction id, adds the specified todo effect to all
      * `TRef` values.
      */
-    final def addTodo(txnId: Long, journal: Journal, todoEffect: Todo): Unit = {
-      val it = journal.entrySet().iterator()
+    final def addTodo(txnId: TxnId, journal: Journal, todoEffect: Todo): Boolean = {
+      var added = false
 
-      while (it.hasNext) {
-        val tref = it.next().getValue().tref
+      journal.forEach { (_, value) =>
+        val tref = value.tref
 
         var loop = true
         while (loop) {
           val oldTodo = tref.todo.get
 
-          val newTodo = oldTodo.updated(txnId, todoEffect)
+          if (!oldTodo.contains(txnId)) {
+            val newTodo = oldTodo.updated(txnId, todoEffect)
+
+            loop = !tref.todo.compareAndSet(oldTodo, newTodo)
 
-          loop = !tref.todo.compareAndSet(oldTodo, newTodo)
+            if (!loop) added = true
+          } else loop = false
         }
       }
+
+      added
     }
 
     /**
      * Runs all the todos.
      */
-    def completeTodos[E, A](io: IO[E, A], journal: Journal, platform: Platform): TryCommit[E, A] = {
+    final def completeTodos[E, A](io: IO[E, A], journal: Journal, platform: Platform): TryCommit[E, A] = {
       val todos = collectTodos(journal)
 
       if (todos.size > 0) platform.executor.submitOrThrow(() => execTodos(todos))
 
       TryCommit.Done(io)
     }
 
-    def tryCommitAsync[E, A](journal: Journal, platform: Platform, stm: STM[E, A], txnId: Long, done: AtomicBoolean)(
+    /**
+     * Performs a diff on the two journals, returning only the "new" entries.
+     */
+    final def diffJournals(oldJournal: Journal, newJournal: Journal): Journal = {
+      val diffJournal = new MutableMap[TRefId, Entry](newJournal.size)
+
+      diffJournal.putAll(newJournal)
+
+      newJournal.forEach { (key, _) =>
+        if (oldJournal.containsKey(key)) {
```

Comment:
Is there no change of different values for the same key?

Reply:
I _think_ that for this purpose, we just care whether or not we knew about the existence of the entry, because if we knew it existed, then regardless of its value, we added it to the `todo` list of the references, and therefore don't need to consider it in the diff. Does that make sense?

Reply:
^^^ @regiskuckaertz

Reply:
ah yeah, I get it 👍